adapters/http-base/src/main/java/org/eclipse/hono/adapter/http/AbstractVertxBasedHttpProtocolAdapter.java

```diff
@@ -1094,25 +1074,49 @@ protected final Future<CommandConsumer> createCommandConsumer(
-                    }
                     // wrap the consumer to be able to create a separate span on closing the consumer (errors should just be logged on that span, not the parent)
-                    return new CommandConsumer() {
+                    final ResponseReadyTracker responseReadyTracker = new ResponseReadyTracker() {
+
+                        private final Future<Void> command = responseReady.future()
+                            .compose(ar -> closeCommandConsumer())
+                            .recover(thr -> closeCommandConsumer().compose(v -> Future.failedFuture(thr)));
+
+                        @Override
+                        public Future<Void> waitForCommand() {
+                            return command;
+                        }
+
                         @Override
-                        public Future<Void> close(final SpanContext spanContext) {
+                        public boolean cancel() {
+                            if (requestProcessed.compareAndSet(false, true)) {
+                                responseReady.complete();
+                                return true;
+                            } else {
+                                return false;
+                            }
+                        }
+
+                        private Future<Void> closeCommandConsumer() {
                             final Span closeConsumerSpan = TracingHelper
-                                    .buildChildSpan(tracer, spanContext, "close command consumer",
+                                    .buildChildSpan(tracer, uploadMessageSpan.context(), "close command consumer",
                                             getTypeName())
                                     .withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_CLIENT)
                                     .start();
                             TracingHelper.setDeviceTags(closeConsumerSpan, tenantObject.getTenantId(), deviceId);
                             return consumer.close(closeConsumerSpan.context())
                                     .onFailure(thr -> TracingHelper.logError(closeConsumerSpan, thr))
-                                    .onComplete(ar -> closeConsumerSpan.finish());
+                                    .onComplete(ar -> closeConsumerSpan.finish())
+                                    .recover(thr -> Future.succeededFuture()); // fail of command close sshal not result in error
                         }
                     };
+
+                    if (!requestProcessed.get()) {
+                        // if the request was not responded already, add a timer for triggering an empty response
+                        addCommandReceptionTimer(ctx, responseReadyTracker, ttdSecs, waitForCommandSpan);
+                        setTtdRequestConnectionCloseHandler(ctx, responseReadyTracker, tenantObject.getTenantId(), deviceId, uploadMessageSpan);
+                    } // otherwise the responseReady has already been closed
```

Comment:
FMPOV it would increase readability to move this block above the `responseReadyTracker` declaration and then return the `new ResponseReadyTracker()` directly.

Reply:
I would like to be possible but block needs already created `responseReadyTracker` in order to pass it to timer and close handler (so they would be able to call cancel).

Reply:
Ok, I see.